src/go/k8s/controllers/node/nodes_controller.go

```diff
@@ -0,0 +1,138 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ resources.NodesLister = &NodeWatcher{}
+
+const (
+	nodeIPsList = "node-list.txt"
+)
+
+// NodeWatcher watch for node changes and updates configmap that
+// are tight to Redpanda clusters with external connectivity.
+// The changes happen whenever new nodes is in ready state or
+// Kubernetes cluster loses node.
+type NodeWatcher struct {
+	client.Client
+	Log	logr.Logger
+	Scheme	*runtime.Scheme
+
+	nodeExternalIPs	sync.Map
+	configMapNames	[]types.NamespacedName
+}
+
+// Reconcile is the function that watches for node changes and updates
+// configmaps that has assigned to Redpanda cluster with external
+// connectivity enabled.
+func (n *NodeWatcher) Reconcile(
+	ctx context.Context, req ctrl.Request,
+) (ctrl.Result, error) {
+	log := n.Log.WithValues("node-watcher", req.NamespacedName)
+
+	var node corev1.Node
+	if err := n.Get(ctx, req.NamespacedName, &node); err != nil {
+		log.Error(err, "Unable to fetch node")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var internalIP, externalIP string
+	for _, addr := range node.Status.Addresses {
+		switch addr.Type {
+		case corev1.NodeInternalIP:
+			internalIP = addr.Address
+		case corev1.NodeExternalIP:
+			externalIP = addr.Address
+		case corev1.NodeExternalDNS, corev1.NodeHostName, corev1.NodeInternalDNS:
+		}
+	}
+
+	n.nodeExternalIPs.Store(internalIP, externalIP)
+
+	for _, name := range n.configMapNames {
+		cm := corev1.ConfigMap{}
+		if err := n.Get(context.Background(), name, &cm); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to retrieve config map: %w", err)
+		}
+
+		cm.Data[nodeIPsList] = n.getNodeList()
+
+		if err := n.Update(context.Background(), &cm); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update config map: %w", err)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (n *NodeWatcher) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Node{}).
+		Complete(n)
+}
+
+// RegisterConfigMap register config map for updates in node list
+func (n *NodeWatcher) RegisterConfigMap(name types.NamespacedName) error {
+	cm := corev1.ConfigMap{}
+	if err := n.Get(context.Background(), name, &cm); err != nil {
+		return fmt.Errorf("failed to retrieve config map: %w", err)
+	}
+
+	cm.Data[nodeIPsList] = n.getNodeList()
+
+	if err := n.Update(context.Background(), &cm); err != nil {
+		return fmt.Errorf("failed to update config map: %w", err)
+	}
+
+	n.configMapNames = append(n.configMapNames, name)
+	return nil
+}
+
+// getNodeList return list of pair that consist of internal and external IP
+// of all nodes.
+func (n *NodeWatcher) getNodeList() string {
```

Comment:
the name `getNodeList` is a bit misleading, as if it does a List against k8s. Instead it builds a string with the kv-pairs of an existing map. Perhaps it could even be a function outside of the struct.